var/spack/repos/builtin/packages/openblas/package.py

```diff
@@ -142,70 +121,99 @@ def check_compilers(self):
                     'OpenBLAS @:0.2.19 does not support OpenMP with clang!'
                 )
 
+    @staticmethod
+    def _microarch_target_args(microarch):
+        # This function is constructed using output of compare-micros.py as of:
+        #   OpenBLAS v0.2.20-1615-gfd2ff271 (1/3/2020)
+        #   Spack    v0.13.2-700-gf7b9592eb (1/3/2020)
+        # and information in OpenBLAS' Makefile.system and Spack's
+        # microarchitectures.json
+
+        family = microarch.family
+        no_openblas = ['a64fx', 'i686', 'k10', 'mic_knl', 'nocona', 'pentium2',
+                       'pentium3', 'pentium4', 'power8le', 'power9le']
+        if microarch.name in no_openblas:
+            # Replace microarch with generic ancestor
+            tty.warn("Microarchitecture '{0}' is unknown to OpenBLAS; using "
+                     "generic '{1}'".format(microarch.name, family.name))
+            microarch = family
```

Comment:
Not quite. As it's coded, any new microarchitecture that's added to both OpenBLAS and Spack that _has the same name in both_ will automatically work. The `no_openblas` list will have to be updated if they implement any in that list (unlikely in most cases; I can't imagine anyone adding a pentium2 specialization), or if a new compiler in Spack is added that has no openblas equivalent.

I thought originally of just adding 2, but it adds a lot more content (the number of "just works" mappings exceeds the number of exceptional cases). The mapping also would be reversed from what is needed in the package file: given a spack target, what openblas target should be used.

But I think you're right, adding available OpenBLAS targets is probably the way to go (possibly in the future trying to parse directly from the staged TargetList.txt?), and instead of some of the mapping and exceptions I do here, just traverse Spack's microarchitecture ancestry until I hit one of the available processors.

Reply:
OK, now we're parsing directly from the staged TargetList and trying to get the best match, using the "dynamic" option if we can't find any common ground.